fmriprep/workflows/confounds.py

```diff
@@ -243,3 +262,201 @@ def reverse_order(inlist):
+    import os
+    import numpy as np
+    from niworkflows.nipype import logging
+
+    # To catch edge cases when there are either no noise or signal components
+    LOGGER = logging.getLogger('workflow')
+
+
+    # Pass in numpy array and column base name to generate headers
+    # modified from add_header_func
+    def aroma_add_header_func(np_arr, col_base, comp_nums):
+        import pandas as pd
+        from sys import version_info
+        PY3 = version_info[0] > 2
+
+        df = pd.DataFrame(np_arr, columns=[str(col_base) + str(index) for index in comp_nums])
+        df.to_csv(str(col_base) + "ica_confounds.tsv",
+                  sep="\t" if PY3 else '\t'.encode(), index=None)
+
+        return os.path.abspath(str(col_base) + "ica_confounds.tsv")
+
+    # Partial regression (close to, but not identical to fsl_regfilt)
+    def calc_residuals(x, y):
+        X = np.column_stack(x + [[0] * len(x[0])])
+        beta_hat = np.linalg.lstsq(X, y)[0]
+        y_hat = np.dot(X, beta_hat)
+        residuals = y - y_hat
+
+        return residuals
+
+    # load the txt files from ICA_AROMA
+    melodic_mix = os.path.join(ica_out_dir, 'melodic.ica/melodic_mix')
+    motion_ics = os.path.join(ica_out_dir, 'classified_motion_ICs.txt')
+
+    # -1 since python lists start at index 0
+    motion_ic_indices = np.loadtxt(motion_ics, dtype=int, delimiter=',') - 1
+    melodic_mix_arr = np.loadtxt(melodic_mix, ndmin=2)
+
+    # Return dummy list of ones if no noise compnents were found
+    if motion_ic_indices.size == 0:
+        LOGGER.warn('WARNING: No noise components were classified')
+        no_noise_arr = np.ones((melodic_mix_arr.shape['0'], 1))
+        aggr_tsv = aroma_add_header_func(no_noise_arr, 'no_noise_aggr_', ['00'])
+        nonaggr_tsv = aroma_add_header_func(no_noise_arr, 'no_noise_nonaggr_', ['00'])
+        aroma_confounds = (aggr_tsv, nonaggr_tsv)
+        return aroma_confounds
+
+    # transpose melodic_mix_arr so x refers to the correct dimension
+    aggr_confounds = np.asarray([melodic_mix_arr.T[x] for x in motion_ic_indices])
+
+    # the "good" ics, (e.g. not motion related)
+    good_ic_arr = np.delete(melodic_mix_arr, motion_ic_indices, 1).T
+
+    # return dummy lists of zeros if no signal components were found
+    if good_ic_arr.size == 0:
+        LOGGER.warn('WARNING: No signal components were classified')
+        no_signal_arr = np.zeros((melodic_mix_arr.shape[0], 1))
+        aggr_tsv = aroma_add_header_func(no_signal_arr, 'no_signal_aggr_', ['00'])
+        nonaggr_tsv = aroma_add_header_func(no_signal_arr, 'no_signal_nonaggr_', ['00'])
+        aroma_confounds = (aggr_tsv, nonaggr_tsv)
+        return aroma_confounds
+
+    # nonaggr denoising confounds
+    nonaggr_confounds = np.asarray([calc_residuals(good_ic_arr, y) for y in aggr_confounds])
+
+    # add one to motion_ic_indices to match melodic report.
+    aggr_tsv = aroma_add_header_func(aggr_confounds.T, 'aggr_comp_',
```

Comment:
All other columns use camel case. Could we do "AROMAAggrComp01" and "AROMANonAggrComp01"?